src/mam4xx/gas_chem.hpp

```diff
@@ -0,0 +1,391 @@
+        converged[kk] = true;
+
+        int mm = permute_4[kk];
+        bool frc_mask = haero::abs(forcing[mm]) > small ? true : false;
+        if (frc_mask) {
+          converged[kk] =
+              haero::abs(forcing[mm]) <= epsilon[kk] * haero::abs(solution[mm])
+                  ? true
+                  : false;
+        } else {
+          converged[kk] = true;
+
+        } // frc_mask
+        if (!converged[kk]) {
+          convergence = false;
+        }
+      } // end
+
+      if (convergence) {
+        return;
+      }
+
+    } // end nr_iter
+
+  } // end nr_iter
+
+} // newton_raphson_iter
+
+KOKKOS_INLINE_FUNCTION
+void imp_sol(Real base_sol[gas_pcnst], //    ! species mixing ratios [vmr] & !
+                                       //    inout
+             const Real reaction_rates[rxntot], const Real het_rates[gas_pcnst],
+             const Real extfrc[extcnt], Real &delt,
+             const int permute_4[gas_pcnst], const int clsmap_4[gas_pcnst],
+             const bool factor[itermax], Real epsilon[clscnt4],
+             Real prod_out[clscnt4], Real loss_out[clscnt4]
+             //                   xhnm, ncol, lchnk, ltrop
+) {
+
+  //     !-----------------------------------------------------------------------
+  // ! ... imp_sol advances the volumetric mixing ratio
+  // ! forward one time step via the fully implicit euler scheme.
+  // ! this source is meant for small l1 cache machines such as
+  // ! the intel pentium and itanium cpus
+  // !-----------------------------------------------------------------------
+  const Real zero = 0;
+  const Real half = 0.5;
+  const Real one = 1;
+  const Real two = 2;
+
+  const int cut_limit = 5;
+
+  Real ind_prd[clscnt4] = {};
+  Real lin_jac[nzcnt] = {};
+  bool converged[clscnt4] = {};
+  bool convergence = false;
+  Real prod[clscnt4] = {};
+  Real loss[clscnt4] = {};
+  Real max_delta[clscnt4] = {};
+  // ! initiate variables
+  // prod_out(:,:,:) = 0._r8
+  // loss_out(:,:,:) = 0._r8
+  // solution(:) = 0._r8
+
+  // !-----------------------------------------------------------------------
+  // ! ... class independent forcing
+  // !-----------------------------------------------------------------------
+  indprd(4,                     //                  & ! in
+         ind_prd,               //               & ! inout
+         reaction_rates, extfrc //   & ! in
+  );                            // ! in
+
+  // level_loop : do lev = 1,pver
+  //    column_loop : do icol = 1,ncol
+  //       if (lev <= ltrop(icol)) cycle column_loop
+
+  // !-----------------------------------------------------------------------
+  // ! ... transfer from base to local work arrays
+  // !-----------------------------------------------------------------------
+
+  // do mm = 1,rxntot
+  //   lrxt(mm) = reaction_rates(icol,lev,mm)
+  // enddo
+  // if( gas_pcnst > 0 ) then
+  //   do mm = 1,gas_pcnst
+  //     lhet(mm) = het_rates(icol,lev,mm)
+  //   enddo
+  // endif
+
```

Comment:
I may be missing something but there is a level loop as well that I don't see here. Is it incorporated in some other way?

Reply:
This implementation will  be called by a kokkos::parallel_for, so we do not need this part of the code in part of the c++ port.

Reply:
Oh yes, I forgot about that. Thanks!